### PR TITLE
TLS: add SSL_writev

### LINKED ISSUE #20918: Add SSL_writev and SSL_writev_ex
I would like to implement `SSL_writev()` and `SSL_writev_ex()` to support writing vectored data in a single record. 

As discussed [here](https://stackoverflow.com/questions/38198638/openssl-ssl-write-from-multiple-buffers-ssl-writev), the current solution would be having the user consolidates the data manually then call `SSL_write()` on the consolidated data.

Therefore, I propose adding two new APIs to combine this process:
```c
int SSL_writev(SSL *s, const struct iovec *iov, int iovcnt);
int SSL_writev_ex(SSL *s, const struct iovec *iov, size_t iovcnt, size_t *written);
```
It will require adding `ssl_writev()` and `ssl_writev_bytes` in `struct ssl_method_st`. `SSL_writev` then can make a call to `ssl_writev_internal` where it calls `ssl3_writev`.

Depending on the negotiated protocols, `ssl3_writev` then may call `ssl3_writev_bytes` or `dtls1_writev_app_data_bytes`.

In `ssl3_writev_bytes`, for each pipelines, we can allocate a temporary buffer with the size being the length of data we want to send through the pipeline. We then fill the buffer with the vectored data, possibly using `memcpy`. By constructing a temporary buffer, we can use the existing `write_records` call.

In `dtls1_writev_app_data_bytes`, we can also allocate a temporary buffer with the size being the total length of the vectored data then copy everything over and the subsequent calls should be the same.

### PR DESCRIPTION
**Pull request by peppapighs**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

Implemented `SSL_writev()` as mentioned in #20918

### CODE REVIEW

File: include/internal/iovec.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    size_t ptr = 0;
+    
+    while (offset >= src[ptr].iov_len) {
+        offset -= src[ptr].iov_len;
+        ptr++;
+    }
+
+    while (len > 0) {
+        size_t to_copy = src[ptr].iov_len - offset;
+        if (to_copy > len)
```
Comment (paulidale) [inline include/internal/iovec.h:39]: blank line above.

File: ssl/record/rec_layer_s3.c [hunk 2]
```diff
@@ -437,6 +437,228 @@ int ssl3_write_bytes(SSL *ssl, int type, const void *buf_, size_t len,
...
+    if (s->s3.alert_dispatch > 0) {
+        i = ssl->method->ssl_dispatch_alert(ssl);
+        if (i <= 0) {
+            /* SSLfatal() already called if appropriate */
+            return i;
+        }
+        /* if it went, fall through and send more stuff */
+    }
+
+    n = (len - tot);
```
Comment (paulidale) [inline ssl/record/rec_layer_s3.c:547]: drop parenthesis?

File: ssl/ssl_lib.c [hunk 3]
```diff
@@ -2506,6 +2510,57 @@ int ssl_write_internal(SSL *s, const void *buf, size_t num, size_t *written)
...
+        args.s = s;
+        args.buf = (void *)iov;
+        args.num = iovcnt;
+        args.type = WRITEVFUNC;
+        args.f.func_writev = s->method->ssl_writev;
+
+        ret = ssl_start_async_job(s, &args, ssl_io_intern);
+        *written = sc->asyncrw;
+        return ret;
+    } else {
```
Comment (paulidale) [inline ssl/ssl_lib.c:2559]: Drop the else clause since it's returning.

**Review by paulidale — COMMENTED** [review]
Style nits only.

File: ssl/record/rec_layer_s3.c (hunk 2 above)
Comment (peppapighs) [reply ssl/record/rec_layer_s3.c:547]: I copied this function from `ssl3_write_bytes` which has parenthesis.

File: ssl/ssl_lib.c (hunk 3 above)
Comment (peppapighs) [reply ssl/ssl_lib.c:2559]: I also copy this function from `ssl_write_internal` where there is an else clause after return.

File: ssl/record/rec_layer_s3.c (hunk 2 above)
Comment (t8m) [reply ssl/record/rec_layer_s3.c:547]: Existing code can have style issues. No need to copy them.

File: ssl/ssl_lib.c (hunk 3 above)
Comment (t8m) [reply ssl/ssl_lib.c:2559]: As above.

**Comment by t8m** [comment]
Rather than using force pushes, please use fixup commits `git commit --fixup <commit-id-to-be-fixed>`

**Comment by peppapighs** [comment]
@t8m Sorry to trouble you but can you enable the workflow again? I had to force push to edit my commit email to solve the CLA issue.

**Comment by peppapighs** [comment]
> Do we need/want both API calls? Seems that just one would be enough since they are very similar.

I try to mirror the SSL_write() and SSL_write_ex() APIs. Realistically, only SSL_writev_ex() should be enough.

**Comment by paulidale** [comment]
Thanks.  I'd lean to calling the new function just `SSL_writev` but other could have a different viewpoint.

File: include/openssl/ssl.h.in [hunk 4]
```diff
@@ -1933,6 +1933,8 @@ __owur ossl_ssize_t SSL_sendfile(SSL *s, int fd, off_t offset, size_t size,
                                  int flags);
 __owur int SSL_write(SSL *ssl, const void *buf, int num);
 __owur int SSL_write_ex(SSL *s, const void *buf, size_t num, size_t *written);
+__owur int SSL_writev_ex(SSL *s, const OSSL_IOVEC *iov, size_t iovcnt,
```
Comment (t8m) [inline include/openssl/ssl.h.in:1936]: This should not be called _ex function if there is no non-ex function.

**Comment by t8m** [comment]
This API won't achieve single copy semantics (at least with QUIC). Adding OTC hold.

**Comment by paulidale** [comment]
Performance figures in: https://github.com/peppapighs/ssl-writev-benchmark/blob/implementation/README.md

**Comment by peppapighs** [comment]
Updated link to performance figures: https://github.com/peppapighs/ssl-writev-benchmark/blob/main/README.md

**Comment by hlandau** [comment]
I've written up my thoughts for our OTC discussion tomorrow.

- iovec-style functions are usually created to avoid excessive copying. However, in this case, SSL_write() could just be called repeatedly. This particular proposal is *not* about avoiding copying but about optimising TLS record size.

- The author of the PR has produced some nice performance benchmarks, but these show that consolidating data using memcpy before calling SSL_write (in order to generate larger TLS records) has comparable performance to that obtained using the SSL_writev() implementation. Thus from a performance perspective there is not anything urgently compelling here.

- There could be other ways to implement this without a new API, for example by allowing libssl to maintain an in-progress unfinished record across multiple SSL_write calls. We would need a suitable API to force flushing; possibly one already exists(?). (@mattcaswell: feasibility?)

  :arrow_right: This IMO is a reason against merging this as is; we may be able to avoid the need for any new APIs.

- This API creates a dependency on `struct iovec`, which is not available on all platforms and thus cannot be merged as is. It would need revision with portability in mind. We have already created our own internal iovec-like structures for internal use with our QUIC code.

  :arrow_right: Cannot be merged as-is.

- A comparable BIO interface would also need to be incorporated. There is a question as to how backwards compatibility should work here.

- The issue of QUIC has been mentioned. There is no implementation issue for QUIC as such but it is likely the performance will be no better than successive SSL_write() calls. We intend to offer another API in future to allow single-copy writing for QUIC, but this will require transfer of buffer ownership back and forth between libssl and the application. But as mentioned above, this PR is not actually about avoiding copying so this is not the issue per se. The issue is more about whether SSL_writev actually offers a compelling advantage that couldn't simply be accomplished with changes to the implementation of SSL_write.

My personal recommendation would be for the investigation of the feasibility of giving applications better control over TLS record generation across successive SSL_write() calls, possibly by allowing a record-in-progress to be maintained. But the application currently has the option of consolidating the
buffer beforehand, and the performance figures the author provides shows this to not make an enormous difference relative to this approach.

While in general I support having interfaces like this I don't think there are sufficiently compelling grounds to adopt it at this time, or before the alternatives have been investigated.

**Comment by t8m** [comment]
**OTC concurs with the above write up.**

**Comment by hlandau** [comment]
This cannot be merged as is, as per the above writeup. This feature might be possible in future and we will discuss this after 3.2; [see comment](https://github.com/openssl/openssl/issues/20918#issuecomment-1588833343) in linked issue.

**Comment by travisdowns** [comment]
> The author of the PR has produced some nice performance benchmarks, but these show that consolidating data using memcpy before calling SSL_write (in order to generate larger TLS records) has comparable performance to that obtained using the SSL_writev() implementation. Thus from a performance perspective there is not anything urgently compelling here.

Yes, but this it's worth noting that applications need to allocate their own ~record size buffer to accomplish this. It's pure waste given OpenSSL immediately copies the user buf into its similarly sized `wbuf` immediately. So it's:

1) A huge perf win if users aren't carefully linearizing their vectors in the application (very common).
2) A small perf win _and significant memory savings per connection_ if they are.

**Pull request closed without merging**